src/Disks/ObjectStorages/DiskObjectStorage.cpp

```diff
@@ -516,16 +521,51 @@ std::unique_ptr<ReadBufferFromFileBase> DiskObjectStorage::readFile(
     std::optional<size_t> read_hint,
     std::optional<size_t> file_size) const
 {
-    auto storage_objects = metadata_storage->getStorageObjects(path);
+    const auto objects = metadata_storage->getStorageObjects(path);
 
     const bool file_can_be_empty = !file_size.has_value() || *file_size == 0;
-
-    if (storage_objects.empty() && file_can_be_empty)
+    if (objects.empty() && file_can_be_empty)
         return std::make_unique<ReadBufferFromEmptyFile>();
 
+    ReadSettings read_settings = object_storage->getAdjustedSettingsFromMetadataFile(updateResourceLink(settings, getReadResourceName()), path);
+    read_settings.enable_page_cache = read_settings.page_cache != nullptr && (enable_page_cache || read_settings.force_enable_page_cache);
+
+    if (read_settings.enable_page_cache)
+    {
+        const bool async_read = settings.remote_fs_method == RemoteFSReadMethod::threadpool;
+
+        ReadSettings modified_settings{read_settings};
+        modified_settings.remote_fs_method = RemoteFSReadMethod::read;
+        std::unique_ptr<ReadBufferFromFileBase> impl = object_storage->readObjects(
+            objects,
+            modified_settings,
+            read_hint,
+            file_size);
+
+        /// We currently don't detect if a file is modified. If we need to use this cache for mutable
+        /// files, we'd have to put some unique identifier in cache_key.file_version.
+        auto cache_key = FileChunkAddress {
+            .disk_name = getName(),
+            .path = path };
+        impl = std::make_unique<CachedInMemoryReadBufferFromFile>(
+            cache_key, read_settings.page_cache, std::move(impl), read_settings);
```

Comment:
As you wrote - it is indeed better to use blob path instead of file path here, it will make things easier. For filesystem cache we add a `CachedOnDiskReadBufferFromFIle` on `CachedObjectStorage` level, so cache is automatically using blob paths for cache keys. May be you'd better move user space page cache there as well? Though it will mean that `CachedObjectStorage` needs to be created not only if fs cache enabled, but also if only user space page cache is enabled without fs cache, which is a complication.